src/ut/utils_test.cpp

```diff
@@ -198,3 +219,527 @@ TEST_F(UtilsTest, ParseIPAddresses)
+  }
+  CW_IO_COMPLETES();
+
+  // Only the first hook gets triggered so x should be 2.
+  EXPECT_EQ(x, 2);
+}
+
+TEST_F(UtilsTest, IOHooksMultipleIOOperations)
+{
+  int count = 0;
+
+  IOHook hook(IOHook::NOOP_ON_START,
+              [&count](const std::string& reason) { count++; });
+
+  CW_IO_STARTS("One") { } CW_IO_COMPLETES();
+  CW_IO_STARTS("Two") { } CW_IO_COMPLETES();
+  CW_IO_STARTS("Three") { } CW_IO_COMPLETES();
+
+  // Only the first hook gets triggered so x should be 2.
+  EXPECT_EQ(count, 3);
+}
+
+TEST_F(UtilsTest, IOHooksArePerThread)
+{
+  int count = 0;
+  std::atomic_bool terminate(false);
+
+  std::thread t1([&count, &terminate] () {
+    IOHook hook(IOHook::NOOP_ON_START,
+                [&count](const std::string& reason) { count++; });
+
+    while (!terminate.load()) { sleep(1); }
+  });
+
+  CW_IO_STARTS("One") { } CW_IO_COMPLETES();
+  CW_IO_STARTS("Two") { } CW_IO_COMPLETES();
+  CW_IO_STARTS("Three") { } CW_IO_COMPLETES();
+
+  // The hook does not get triggered so the count should be 0
+  EXPECT_EQ(count, 0);
+
+  terminate.store(true);
+  t1.join();
+}
+
+class StopWatchTest : public ::testing::Test
+{
+public:
+  StopWatchTest()
+  {
+    cwtest_completely_control_time();
+  }
+
+  virtual ~StopWatchTest()
+  {
+    cwtest_reset_time();
+  }
+
+  unsigned long ms_to_us(int ms) { return (unsigned long)(ms * 1000); }
+
+  Utils::StopWatch _sw;
+};
+
+TEST_F(StopWatchTest, Mainline)
+{
+  EXPECT_TRUE(_sw.start());
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.stop());
+
+  unsigned long elapsed_us;
+  EXPECT_TRUE(_sw.read(elapsed_us));
+  EXPECT_EQ(ms_to_us(11), elapsed_us);
+}
+
+TEST_F(StopWatchTest, StopIsIdempotent)
+{
+  EXPECT_TRUE(_sw.start());
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.stop());
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.stop());
+
+  unsigned long elapsed_us;
+  EXPECT_TRUE(_sw.read(elapsed_us));
+  EXPECT_EQ(ms_to_us(11), elapsed_us);
+}
+
+TEST_F(StopWatchTest, ReadGetsLatestValueWhenNotStopped)
+{
+  EXPECT_TRUE(_sw.start());
+
+  unsigned long elapsed_us;
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.read(elapsed_us));
+  EXPECT_EQ(ms_to_us(11), elapsed_us);
+
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.read(elapsed_us));
+  // The returned value is greater on the second read.
+  EXPECT_EQ(ms_to_us(22), elapsed_us);
+}
+
+TEST_F(StopWatchTest, StopStartNotIncludedInReading)
+{
+  EXPECT_TRUE(_sw.start());
+  cwtest_advance_time_ms(11);
+
+  // Stop and restart the stopwatch a few times. The time increases in between
+  // should not be included in the final value read.
+  EXPECT_TRUE(_sw.stop());
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.start());
+
+  EXPECT_TRUE(_sw.stop());
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.start());
+
+  cwtest_advance_time_ms(11);
+  EXPECT_TRUE(_sw.stop());
+
+  unsigned long elapsed_us;
+  EXPECT_TRUE(_sw.read(elapsed_us));
+  EXPECT_EQ(ms_to_us(22), elapsed_us);
+}
+
+TEST_F(StopWatchTest, StopStartThenRead)
```

Comment:
Isn't this test the same as the one above it?

Reply:
Not quite. I the second test the stopwatch is running when the read happens. I have clarified by making the test name more descriptive and adding come more comments.